### DISCUSSION TITLE
Trying to record a video, getting an error

### DISCUSSION THREAD

**Discussion by antongenis** [bluenviron/mediamtx#634] [Questions]
### Discussed in https://github.com/aler9/rtsp-simple-server/discussions/598

<div type='discussions-op-text'>

<sup>Originally posted by **antongenis** September 29, 2021</sup>
2021/09/29 11:26:18 I [1/0] [RTSP] [session 644619446] opened by 127.0.0.1:38766
[rtsp @ 0x5555c885b900] method SETUP failed: 461 Unsupported Transport**
2021/09/29 11:26:18 I [1/0] [RTSP] [session 158090036] opened by 127.0.0.1:38766

the command I am using to send a video:

ffmpeg -re -i ipd.mp4 -c:v libx264 -preset superfast -tune zerolatency -f rtsp -rtsp_transport tcp rtsp://SS:DD@52.233.254.0:8554/stream7    
___________________________________________________________________________________________
in the yml file:

runOnPublish: ffmpeg -i rtsp://localhost:$RTSP_PORT/$RTSP_PATH -rtsp_transport tcp -c:v libx264 /mnt/video-dev/$RTSP_PATH.mp4

</div>

**Comment by antongenis** [comment]
I am seeing that the file is being recorded, but not completely, not all frames are there

the original video has 247 frames , and the one recorded -220

$ ffprobe -select_streams v -show_streams stream7.mp4 2>/dev/null | grep nb_frames | sed -e 's/nb_frames=//'
220

**Comment by aler9** [comment]
An H264 stream contains frames encoded as I-frames, B-frames and P-frames. B-frames and P-frames are partial frames, they contain only the modified portion of the stream and can't be decoded autonomously. In your case, this happens:

1. You start streaming a file, this get sent to the server (I = i-frame, P = p-frame):

    ```
   I P P P P P P P P P P P I P P P P P P P P P P P I
    ```

2. The server starts the runOnPublish command, but some milliseconds are passed, and this is what the ffmpeg inside the command receives:

   ```
   P P P P P I P P P P P P P P P P P I
   ```
   
   The first I-frame has been transmitted before the runOnPublish ffmpeg is able to connect, therefore it is not recorded. At the same time, the subsequent P-frames can't be recorded, since the first I-frame has not been recorded and they can't be decoded. The runOnPublish ffmpeg can record only from the 2nd I-frame onward.
   
This is reason behind the missing frames.

An enhancement could be adding a record function directly inside the server, that is able to record all frames without the initial delay of a runOnPublish command, but at the moment this is not implemented.

[a comment by an automated account is not shown]
